### DISCUSSION TITLE
[General] Paymaster transaction reverts only on token approval transactions. Seems it's not a contract error.

### DISCUSSION THREAD

**Discussion by velocore** [zkSync-Community-Hub/zksync-developers#316] [General] [answered]
### Environment

zkSyncEra Mainnet

### Provide a brief description of the functionality you're trying to implement.

Send transaction using ERC20 as gas fee through Paymaster.

### What is the specific issue or error you're encountering?

It works perfectly on all other txs and worked perfectly until Monday, but suddenly it reverts only to token approval transactions.
Swap, add liquidity, or other transactions are working well.

### Can you share the error messages or logs you're receiving, if any?

This is an example tx reported by the user, and everyone is facing the same error.
https://explorer.zksync.io/tx/0x4b1ded9c6e86daf88eaf2aca53f28b29c93badeba1cd24e1660af08a49dd3ac3
[Error: function_selector = 0x, data = 0x]
If you trace the log, execution itself was done till the end but eventually reverted at the end without any reason. So it seems this isn't a solidity-side error. All other transactions had no problems.

### Have you made any recent changes to the contract before encountering this issue?

Nope. we are using same verified paymaster here : 
https://explorer.zksync.io/address/0x86E6Ab1950770e24035D1b3a5907C7D503d43B01

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?

It was executed without problem till the end and it wasn't a out of gas error.

### Have you tried to isolate the problem, and if so, what were the results?

Compared to other protocol's paymaster call data, such as syncswap, we didn't find any differences.

### What steps have you already taken to try to resolve the issue?

When it first happened last week, we were using Ankr's premium RPC. Changing to the public RPC for transaction sending immediately fixed the issue. However, when it happened again this week, changing RPC didn't solve anything. Both the public and Ankr's RPC are facing the same issue.

### Additional Details

Attaching transaction sign data. You could reproduce this by just making any ERC20 token swap with the paymaster option on. I will get this error on the approval step.

```
TxType:113
From:194942874177707172163702468349261938307471787296
To:640729936092112396866838086980521122890860456039
GasLimit:441641
GasPerPubdataByteLimit:50000
MaxFeePerGas:100000000
MaxPriorityFeePerGas:0
Paymaster:770148839500966453391380883766730123028578188033
Nonce:154
Value:0
Data:0x095ea7b3000000000000000000000000f5e67261cb357edb6c7719fefafaab280cb5e2a6ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
FactoryDeps:
PaymasterInput:
0x949431dc000000000000000000000000ed4040fd47629e7c8fbb7da76bb50b3e7695f0f20000000000000000000000000000000000000000000000000050a1b81c48524400000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000
```

**Comment by lutr0** [comment]
We are having the same exact issue at Zyfi, also since Monday without any changes from our side.

Most transactions are not even submitted on chain but fail after signature, example:

```
An internal error was received.

Request Arguments:
  chain:         zkSync Era (id: 324)
  from:          0x79a4c883f7654ea81e17848c8acbfece7d9234d2
  to:            0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4
  value:         0 ETH
  data:          0x095ea7b30000000000000000000000004bba932e9792a2b917d47830c93a9bc79320e4f700000000000000000000000000000000000000000000000000000000001e8480
  gas:           695547
  maxFeePerGas:  0.1 gwei

Details: failed to validate the transaction. reason: Validation revert: Paymaster validation error: Error function_selector = 0x0757dbc9, data = 0x0757dbc9
Version: viem@2.7.6
```

Sometimes instead the error is about a low level fail for safeERC20

**Comment by dutterbutter** [comment]
Replaying the transaction [`0x4b1ded9c6e86daf88eaf2aca53f28b29c93badeba1cd24e1660af08a49dd3ac3`](https://explorer.zksync.io/tx/0x4b1ded9c6e86daf88eaf2aca53f28b29c93badeba1cd24e1660af08a49dd3ac3) using `era-test-node` yields some interesting / more details (summarized):

```bash
Call(Normal) 0x66621cc5df42d1a2dcb78aba257ea8d945b67df3 executeTransaction(bytes32, bytes32, tuple) Revert: Unknown revert reason  1008882
19:11:23  INFO Call(Normal) 0xd092e42453d6864ea98597461c50190e372d2448 mint(bytes)  Error: Panic 987147

Call(Normal) System context setTxOrigin(address)   1964277
19:11:49  INFO Call(Normal) 0xf6867617fca74aff62e30a897e7fc372ed0cfa7a executeTransaction(bytes32, bytes32, tuple) Revert: Unknown revert reason  1959552
19:11:49  INFO   Call(Normal) 0xf8b59f3c3ab33200ec80a8a58b2aa5f5d2a8944c multicall(uint256,bytes[]) Revert: Unknown revert reason  1901214
19:11:49  INFO     Call(Normal) System context getBlockTimestamp()   1870659
19:11:49  INFO     Call(Delegate) 0xf8b59f3c3ab33200ec80a8a58b2aa5f5d2a8944c exactInputSingle((address,address,uint24,address,uint256,uint256,uint160)) Revert: Unknown revert reason  1868958
19:11:49  INFO       Call(Normal) ImmutableSimulator getImmutable(address, uint256)   1803438
19:11:49  INFO         Call(Normal) Keccak 00000000   1774710
19:11:49  INFO         Call(Normal) Keccak 00000000   1774206
19:11:49  INFO       Call(Normal) Keccak 00000000   1801296
19:11:49  INFO       Call(Normal) Keccak 2020dba9   1800666
19:11:49  INFO       Call(Normal) Account Code Storage getCodeSize(uint256)   1798524
19:11:49  INFO       Call(Normal) 0x291d9f9764c72c9ba6ff47b451a9f7885ebf9977 swap(address,bool,int256,uint160,bytes) Revert: Unknown revert reason  1776978
```

I'll need to do some further investigating, possibly in collaboration with the protocol team. If you're interested in replaying the transaction, make sure to pull the latest from the `era-test-node` main branch or from latest release `v0.1.0-alpha.16`:

```bash
./target/release/era_test_node --show-calls=all --resolve-hashes replay_tx mainnet 0x4b1ded9c6e86daf88eaf2aca53f28b29c93badeba1cd24e1660af08a49dd3ac3
```

**Reply by lutr0** [reply]
Thank you for looking into it!
To give you some additional (potentially useful) data:

Trying to send an approval transaction like this one (that was working before monday), fails before the tx is sent:
`failed to validate the transaction. reason: Validation revert: Paymaster validation error: Error function_selector = 0x0757dbc9, data = 0x0757dbc9`

```
{
    "rawTx": {
        "from": "0x79A4c883f7654ea81E17848c8aCbfeCe7d9234d2",
        "to": "0x3355df6d4c9c3035724fd0e3914de96a5a83aaf4",
        "data": "0x095ea7b30000000000000000000000004bba932e9792a2b917d47830c93a9bc79320e4f700000000000000000000000000000000000000000000000000000000001e8480",
        "value": "0",
        "customData": {
            "paymasterParams": {
                "paymaster": "0xa2D1AE31C2236fBE054c6C9fBDEF27e78ac78D71",
                "paymasterInput": "0x949431dc0000000000000000000000003355df6d4c9c3035724fd0e3914de96a5a83aaf4000000000000000000000000000000000000000000000000000000000000cf9c000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000018d8a288f1b0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000004124210cfc189bb38d932cc3dfa975b84c50fad69a77c5ef9faea1ca0116fe1ad2494b1ac9e7470511774bc47aefb0314427c2aeb768e17be26cf95cb634282a421c00000000000000000000000000000000000000000000000000000000000000"
            },
            "gasPerPubdata": 50000
        },
        "maxFeePerGas": "100000000",
        "gasLimit": 2190900
    },
    "networkId": 324
}
```

**Reply by MexicanAce** [reply]
Okay, got more information from `era_test_node` regarding the paymaster failure. Here's some interesting call stack trace lines:
1. `Call(Normal) Account Code Storage                                 getRawCodeHash(address) Revert: Error function_selector = 0x0908c379, data = 0x0908c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000205361666545524332303a206c6f772d6c6576656c2063616c6c206661696c6564`
2. `Call(Normal) 0x86e6ab1950770e24035d1b3a5907c7d503d43b01                 0x038a24bc Revert: SafeERC20: low-level call failed`
3. `Call(Normal) 0xed4040fd47629e7c8fbb7da76bb50b3e7695f0f2           transferFrom(address,address,uint256)  Error: Panic`

The error `SafeERC20: low-level call failed` is coming from here: https://github.com/matter-labs/era-contracts/blob/main/system-contracts/contracts/openzeppelin/token/ERC20/utils/SafeERC20.sol#L141

**Reply by lutr0** [reply, marked as answer]
From local testing, the issue seems to be solved by assigning a higher `gasLimit`.

**Comment by tus4ncockverb** [comment]
The issue seems to be resolved by assigning a higher gasLimit.

**Comment by quanteauphil** [comment]
It seems like the issue you're encountering with ERC20 token approval transactions through the Paymaster is not related to gas estimation or RPC provider. Since the execution completes successfully but the transaction eventually reverts without any reason, it could be an issue with the Paymaster contract itself or the interaction between your contract and the Paymaster. I would recommend thoroughly reviewing the Paymaster contract code and ensuring that all necessary conditions for ERC20 approval are met. Additionally, consider reaching out to the zkSync team for further assistance in debugging this issue.

**Comment by mcgeesusan** [comment]
It seems like there may be an issue with the Paymaster contract or its interaction with the zkSyncEra mainnet. I recommend reaching out to the zkSync team for further assistance and debugging of this issue.

**Comment by matthewredmanv** [comment]
It appears that the issue may be related to the transaction data or the interaction with the paymaster contract. Please review the transaction data and the paymaster contract to ensure compatibility and correct functioning. Additionally, consider reaching out to the zkSync support team for further assistance in troubleshooting the issue.

**Comment by glasrudmary** [comment]
It seems that there is an issue with sending ERC20 tokens as gas fee through the Paymaster on zkSyncEra Mainnet. The transactions are reverting only for token approval transactions, while other transactions are working well. The error message shows function_selector = 0x, data = 0x. The issue has been reported by multiple users and changing the RPC did not solve the problem. Further investigation and debugging may be required to resolve this issue.
